app/models/concerns/scholars_archive/date_operations.rb

```diff
@@ -0,0 +1,134 @@
+module ScholarsArchive
+  module DateOperations
+    extend ActiveSupport::Concern
+
+    def to_solr(solr_doc={})
+      solr_doc = super
+      solr_doc = solr_doc.merge({"date_decades_ssim" => decades})
+      solr_doc = solr_doc.merge({"date_facet_yearly_ssim" => date_facet_yearly})
+      solr_doc
+    end
+
+    def date_facet_yearly
+      return nil if yearly_dates.empty?
+      yearly_dates
+    rescue ArgumentError
+    end
+
+    def decades
+      return nil if decade_dates.empty?
+      decade_dates.map(&:decade)
+    rescue ArgumentError
+    end
+
+    # Determine the date value to use for Decades facet and date facet yearly processing.
+    def date_value
+      if date_created.present? then
+        date_created
+      elsif date_copyright.present? then
+        date_copyright
+      elsif date_issued.present? then
+        date_issued
+      end
+    end
+
+    def yearly_dates
+      (date_value) ? clean_years : []
+    end
+
+    def clean_years
+      date = Date.edtf(date_value)
+      if date.instance_of? EDTF::Interval
+        date.map(&:year).uniq
+      elsif date.instance_of? Date
+        Array.wrap(date.year)
+      else
+        Array.wrap(clean_datetime.year)
+      end
+    end
+
+    def decade_dates
+      dates = DateDecadeConverter.new(date_value).run
+      dates ||= Array.wrap(DecadeDecorator.new(clean_datetime.year))
+    end
+
+    class DecadeDecorator
+      attr_accessor :year
+      def initialize(year)
+        @year = year
+      end
+
+      def decade
+        "#{first_year}-#{last_year}"
+      end
+
+      private
+
+      def first_year
+        year - year%10
+      end
+
+      def last_year
+        year + 10 - (year+10)%10 - 1
+      end
+    end
+
+    class DateDecadeConverter
+      attr_accessor :date
+      def initialize(date)
```

Comment:
It would be nice to have a comment to show expected examples of the value of `date` to help understand how the other related methods are processing.